File: src/drivecatalog/api/models/drive.py
"""Pydantic models for drive-related API responses."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class DriveResponse(BaseModel):
    """Response model for a single drive."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    uuid: str | None
    mount_path: str
    total_bytes: int
    last_scan: datetime | None
    file_count: int
    disk_uuid: str | None = None
    device_serial: str | None = None
    fs_fingerprint: str | None = None


class DriveListResponse(BaseModel):
    """Response model for listing drives."""

    drives: list[DriveResponse]
    total: int


class DriveCreateRequest(BaseModel):
    """Request model for creating/registering a drive."""

    path: str
    name: str | None = None


class DriveRecognizeResponse(BaseModel):
    """Response model for drive recognition."""

    status: str  # recognized, not_found, ambiguous, weak_match
    confidence: str  # certain, probable, ambiguous, weak, none
    drive: DriveResponse | None = None
    candidates: list[DriveResponse] | None = None
    mount_path: str | None = None


class DriveStatusResponse(BaseModel):
    """Response model for drive status."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    mounted: bool
    file_count: int
    folder_count: int
    hashed_count: int
    hash_coverage_percent: float
    last_scan: datetime | None
    first_seen: datetime | None
    video_count: int
    image_count: int
    audio_count: int
    # Disk usage (persisted — available even when disconnected)
    used_bytes: int | None = None
    # Drive health
    smart_status: str | None = None  # "Verified", "Failing", "Not Supported"
    media_type: str | None = None  # "SSD", "HDD"
    device_protocol: str | None = None  # "USB", "SATA", etc.

File: DriveCatalog/Models/Drive.swift
import Foundation

/// Response model for a single drive.
struct DriveResponse: Codable, Identifiable, Hashable {
    let id: Int
    let name: String
    let uuid: String?
    let mountPath: String
    let totalBytes: Int64
    let lastScan: Date?
    let fileCount: Int
    let diskUuid: String?
    let deviceSerial: String?
    let fsFingerprint: String?

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case uuid
        case mountPath = "mount_path"
        case totalBytes = "total_bytes"
        case lastScan = "last_scan"
        case fileCount = "file_count"
        case diskUuid = "disk_uuid"
        case deviceSerial = "device_serial"
        case fsFingerprint = "fs_fingerprint"
    }
}

/// Response model for drive recognition.
struct DriveRecognizeResponse: Codable {
    let status: String  // recognized, not_found, ambiguous, weak_match
    let confidence: String  // certain, probable, ambiguous, weak, none
    let drive: DriveResponse?
    let candidates: [DriveResponse]?
    let mountPath: String?

    enum CodingKeys: String, CodingKey {
        case status
        case confidence
        case drive
        case candidates
        case mountPath = "mount_path"
    }
}

/// Response model for listing drives.
struct DriveListResponse: Codable {
    let drives: [DriveResponse]
    let total: Int
}

/// Request model for creating/registering a drive.
struct DriveCreateRequest: Codable {
    let path: String
    let name: String?
}

/// Response model for drive status.
struct DriveStatusResponse: Codable, Identifiable {
    let id: Int
    let name: String
    let mounted: Bool
    let fileCount: Int
    let folderCount: Int
    let hashedCount: Int
    let hashCoveragePercent: Double
    let lastScan: Date?
    let firstSeen: Date?
    let videoCount: Int
    let imageCount: Int
    let audioCount: Int
    // Disk usage (persisted — available even when disconnected)
    let usedBytes: Int64?
    // Drive health
    let smartStatus: String?
    let mediaType: String?
    let deviceProtocol: String?

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case mounted
        case fileCount = "file_count"
        case folderCount = "folder_count"
        case hashedCount = "hashed_count"
        case hashCoveragePercent = "hash_coverage_percent"
        case lastScan = "last_scan"
        case firstSeen = "first_seen"
        case videoCount = "video_count"
        case imageCount = "image_count"
        case audioCount = "audio_count"
        case usedBytes = "used_bytes"
        case smartStatus = "smart_status"
        case mediaType = "media_type"
        case deviceProtocol = "device_protocol"
    }
}
